File: dsa-learning/dsa-python-code/com/dsa/leetcode/linked_list/P141_LinkedListCycle.py
# author: tobebetter9527
# since: 2023/10/31 10:37
from typing import Optional

from com.dsa.leetcode.linked_list import ListNode


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head:
            return False
        fast = head.next
        slow = head
        while fast and fast.next and fast != slow:
            fast = fast.next.next
            slow = slow.next
        return fast == slow

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/linked_list/two_pointers/P141_LinkedListCycle.java
package com.dsa.leetcode.labuladong.linked_list.two_pointers;

import com.dsa.leetcode.labuladong.linked_list.ListNode;

import java.util.HashSet;
import java.util.Set;

public class P141_LinkedListCycle {
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode fast = head.next, slow = head;
        while (fast != null && fast.next != null && fast != slow) {
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow == fast;
    }

    public boolean hasCycle2(ListNode head) {
        if (head == null) {
            return false;
        }

        Set<ListNode> set = new HashSet<>();
        while (head != null) {
            if (!set.add(head)) {
                return true;
            }
            head = head.next;
        }
        return false;
    }
}
